### PR TITLE
ROX-20184: deduper state pagination

### PR DESCRIPTION
**Pull request by lvalerom**
## Description

If the `DeduperState` has more than `ROX_MAX_DEDUPER_ENTRIES_PER_MESSAGE` (default 200k) entries Central sends the `DeduperState` in chunks of a maximum of `ROX_MAX_DEDUPER_ENTRIES_PER_MESSAGE` entries per chunk. The `DeduperState` has two new fields:
* `Total`: Indicates the total number of chunks that will be sent.
* `Current`: Indicates  from 1 to `Total` which chunk was sent.

Sensor expects the chunks sent in order. If Sensor does not receive the chunks in order it restarts the connection with Central and defaults to not doing reconciliation in Sensor.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

* [x] CI
* [x] Unit tests added
* [x] Manual tests
  * Deploy ACS in a cluster
  * Set ROX_MAX_DEDUPER_ENTRIES_PER_MESSAGE=10: `ks set env deploy/central ROX_MAX_DEDUPER_ENTRIES_PER_MESSAGE=10`
  * On reconnect you should see the deduper chunks are being sent in central and sensor's logs:
```
# Central logs: after reconnect
sensor/service/connection: 2023/10/20 09:57:42.495441 connection_impl.go:605: Info: Sensor (6cfe4d44-48d0-498a-a0a3-a0c39c832d6a) can do client reconciliation: sending deduper state
sensor/service/connection: 2023/10/20 09:57:42.495760 connection_impl.go:704: Info: Sending 10 hashes (Size=603), current chunk: 1, total: 53
sensor/service/connection: 2023/10/20 09:57:42.496014 connection_impl.go:704: Info: Sending 10 hashes (Size=587), current chunk: 2, total: 53
sensor/service/connection: 2023/10/20 09:57:42.496197 connection_impl.go:704: Info: Sending 10 hashes (Size=607), current chunk: 3, total: 53
sensor/service/connection: 2023/10/20 09:57:42.496398 connection_impl.go:704: Info: Sending 10 hashes (Size=597), current chunk: 4, total: 53
sensor/service/connection: 2023/10/20 09:57:42.496551 connection_impl.go:704: Info: Sending 10 hashes (Size=605), current chunk: 5, total: 53
[...]
sensor/service/connection: 2023/10/20 09:57:42.507213 connection_impl.go:704: Info: Sending 10 hashes (Size=591), current chunk: 50, total: 53
sensor/service/connection: 2023/10/20 09:57:42.507590 connection_impl.go:704: Info: Sending 10 hashes (Size=603), current chunk: 51, total: 53
sensor/service/connection: 2023/10/20 09:57:42.508034 connection_impl.go:704: Info: Sending 10 hashes (Size=597), current chunk: 52, total: 53
sensor/service/connection: 2023/10/20 09:57:42.508440 connection_impl.go:704: Info: Sending 4 hashes (Size=235), current chunk: 53, total: 53
sensor/service/connection: 2023/10/20 09:57:42.508743 connection_impl.go:639: Info: Successfully sent deduper state to sensor (6cfe4d44-48d0-498a-a0a3-a0c39c832d6a)
```
```
# Sensor logs: after reconnect
common/sensor: 2023/10/20 09:57:42.564071 central_communication_impl.go:277: Info: Waiting for deduper state from Central
common/sensor: 2023/10/20 09:57:42.564264 central_communication_impl.go:311: Info: Received 10 hashes (size=603), current chunk: 1, total: 53
common/sensor: 2023/10/20 09:57:42.564378 central_communication_impl.go:311: Info: Received 10 hashes (size=587), current chunk: 2, total: 53
common/sensor: 2023/10/20 09:57:42.564496 central_communication_impl.go:311: Info: Received 10 hashes (size=607), current chunk: 3, total: 53
[...]
common/sensor: 2023/10/20 09:57:42.569448 central_communication_impl.go:311: Info: Received 10 hashes (size=591), current chunk: 50, total: 53
common/sensor: 2023/10/20 09:57:42.569540 central_communication_impl.go:311: Info: Received 10 hashes (size=603), current chunk: 51, total: 53
common/sensor: 2023/10/20 09:57:42.569639 central_communication_impl.go:311: Info: Received 10 hashes (size=597), current chunk: 52, total: 53
common/sensor: 2023/10/20 09:57:42.569715 central_communication_impl.go:311: Info: Received 4 hashes (size=235), current chunk: 53, total: 53
common/sensor: 2023/10/20 09:57:42.570104 central_communication_impl.go:188: Info: Established connection to Central.
common/sensor: 2023/10/20 09:57:42.570141 central_communication_impl.go:198: Info: Communication with central started.
```

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: central/sensor/service/connection/connection_impl.go [hunk 1]
```diff
@@ -603,19 +605,37 @@ func (c *sensorConnection) Run(ctx context.Context, server central.SensorService
...
+			if err := c.sendDeduperState(server, successfulHashes, 1, 1); err != nil {
+				return err
+			}
+		}
+		// Since the environment variable can be set to zero or some negative number.
+		// Make sure maxEntries is a valid number to not panic
+		if maxEntries < 1 {
+			maxEntries = 200000
+		}
+		log.Debugf("DeduperState chunk size %d", maxEntries)
```
Comment (fredrb) [inline central/sensor/service/connection/connection_impl.go:621]: nit: move this log one line below and include values of `len(successfulHashes)` and `total`

**Review by fredrb — APPROVED** [review]
Tested this locally and it looks great. Thanks!

File: sensor/common/sensor/central_communication_impl.go [hunk 2]
```diff
@@ -273,34 +275,50 @@ func (s *centralCommunicationImpl) initialDeduperSync(stream central.SensorServi
...
-	case <-time.After(s.syncTimeout):
-		return errTimeoutWaitingForDeduperState
-	case <-done:
-		break
-	}
-	if err != nil {
-		if e, ok := status.FromError(err); ok {
-			if e.Code() == codes.ResourceExhausted {
-				return errors.Wrap(errLargePayload, e.String())
+	var current int32 = 1
```
[the remark this answers is not in the document]
Comment (lvalerom) [reply sensor/common/sensor/central_communication_impl.go]: Done

File: pkg/env/max_deduper_entries_per_message.go [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package env
+
+var (
+	// MaxDeduperEntriesPerMessage sets the max number of deduper entries per message to be sent during the sync
+	MaxDeduperEntriesPerMessage = RegisterIntegerSetting("ROX_MAX_DEDUPER_ENTRIES_PER_MESSAGE", 200000)
+)
```
[the remark this answers is not in the document]
Comment (lvalerom) [reply pkg/env/max_deduper_entries_per_message.go:10]: Done

File: central/sensor/service/connection/connection_impl.go (hunk 1 above)
Comment (lvalerom) [reply central/sensor/service/connection/connection_impl.go]: Done

File: central/sensor/service/connection/connection_impl.go [hunk 4]
```diff
@@ -603,19 +605,37 @@ func (c *sensorConnection) Run(ctx context.Context, server central.SensorService
...
-			return errors.Wrap(err, "unable to sync deduper state")
+		// If there are no hashes we send the empty map
+		if len(successfulHashes) == 0 {
+			if err := c.sendDeduperState(server, successfulHashes, 1, 1); err != nil {
+				return err
+			}
+		}
+		// Since the environment variable can be set to zero or some negative number.
+		// Make sure maxEntries is a valid number to not panic
+		if maxEntries < 1 {
```
[the remark this answers is not in the document]
Comment (lvalerom) [reply central/sensor/service/connection/connection_impl.go]: Done

**Pull request merged**